### PR TITLE
Fixes #802 , Show error Message for all duplicate method overload

### LINKED ISSUE #802: Duplicate identifier error does not show up for method signatures
``` TypeScript
interface A {
    a();     
    a();     
    a();     
}            
```

Expected:

```
file.ts(2,5): error TSxxxx: Duplicate overload signature for 'a'.
file.ts(3,5): error TSxxxx: Duplicate overload signature for 'a'.
file.ts(4,5): error TSxxxx: Duplicate overload signature for 'a'.
```

Actual:

```
```

(no error)

---

Apparently the following also gives no error:

``` TypeScript
function f();
function f();
function f() {
}
```

### PR DESCRIPTION
**Pull request by dekajp**
Fixes #802
error Messages , could be more straight forward for each Type Nodes.
```
  fn IsTypeNodeIdentical() - could be fit inside IsTypeIdentical by comparing TypeFlags 
  , current baselines diff were to much to figure out if this change break something.
```

### CODE REVIEW

**Comment by dekajp** [comment]
@CyrusNajmabadi  i think this could be the problem  `getDeclaredTypeOfTypeParameter` for overload constraint  , Constraint Type is missing .
https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts#L2124

this should be something like this.

```
var constraint = (<TypeParameterDeclaration>getDeclarationOfKind(symbol, SyntaxKind.TypeParameter)).constraint
                if (constraint) {
                    type.constraint = getTypeFromTypeNode(constraint);
                }
                else {
                    type.constraint = noConstraintType;
                }
```

**Comment by dekajp** [comment]
@CyrusNajmabadi   CR feedback implemented . Commit 9d9501b is fixing the error messages reported by commit b432f1c.

**Comment by mhegazy** [comment]
@dekajp sorry for the delay. Just recovering from the holidays :) can you merge master in to get it up-to-date and make sure tests pass.

@CyrusNajmabadi any more comments, or we are ready to go?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2127,7 +2127,10 @@ module ts {
             if (!links.declaredType) {
                 var type = <TypeParameter>createType(TypeFlags.TypeParameter);
                 type.symbol = symbol;
-                if (!(<TypeParameterDeclaration>getDeclarationOfKind(symbol, SyntaxKind.TypeParameter)).constraint) {
+                var constraint = (<TypeParameterDeclaration>getDeclarationOfKind(symbol, SyntaxKind.TypeParameter)).constraint;
+                if (constraint) {
+                    type.constraint = getTypeFromTypeNode(constraint);
+                }else{
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: newline before the 'else' keyword.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2896,10 +2903,14 @@ module ts {
                             }
                         }
                         else {
-                            if (node.typeArguments) {
+                            if (type.flags & (TypeFlags.Class | TypeFlags.Interface) && getNodeLinks(node).isIllegalTypeReferenceInConstraint){
+                                error(node, Diagnostics.Illegal_reference_0_in_constraint, typeToString(type));
+                                type = undefined;
+                            }else if (node.typeArguments) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: newline  before the 'else' keyword

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -2896,10 +2903,14 @@ module ts {
                             }
                         }
                         else {
-                            if (node.typeArguments) {
+                            if (type.flags & (TypeFlags.Class | TypeFlags.Interface) && getNodeLinks(node).isIllegalTypeReferenceInConstraint){
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Space after `)`.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -7219,6 +7227,10 @@ module ts {
             }
 
             checkSpecializedSignatureDeclaration(node);
+            var nodeSymbol = getSymbolOfNode(node);
+            if (nodeSymbol){
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `) {`

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -7515,6 +7527,54 @@ module ts {
...
+            
+            var declarations = symbol.declarations;
+            var duplicateDeclarationIndexes: Map<boolean> = {};
+            var signatures = getSignaturesOfSymbol(symbol);
+
+            for (var i = 0; i < signatures.length; i++) {
+                var signature1 = signatures[i];
+                for (var j = 0; j < i; j++) {
+                    var signature2 = signatures[j];
+                    switch(signature1.declaration.kind){
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `) {`

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use `isAnyFunction`

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -7515,6 +7527,54 @@ module ts {
             return flags & flagsToCheck;
         }
 
+        function checkDuplicateDeclarationOfSymbolInSameBlock(symbol: Symbol): void {
+            if ( !symbol.declarations) return;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Remove space before `!`

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: When would a symbol not have declarations?

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I think what you want to do is check whether or not there's more than 1 declaration. If there is only 1, bail out.

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -7515,6 +7527,54 @@ module ts {
...
+                    var signature2 = signatures[j];
+                    switch(signature1.declaration.kind){
+                        case SyntaxKind.FunctionDeclaration:
+                        case SyntaxKind.MethodDeclaration:
+                        case SyntaxKind.MethodSignature:
+                        case SyntaxKind.Constructor:
+                        case SyntaxKind.CallSignature:
+                        case SyntaxKind.ConstructSignature:
+                            if (signature1.declaration.parent === signature2.declaration.parent && 
+                                compareSignatures(signature1, signature2, /*compareReturnTypes*/ false, compareTypes)=== Ternary.True) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Space before `===`

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -7515,6 +7527,54 @@ module ts {
...
+        function checkDuplicateDeclarationOfSymbolInSameBlock(symbol: Symbol): void {
+            if ( !symbol.declarations) return;
+            
+            var declarations = symbol.declarations;
+            var duplicateDeclarationIndexes: Map<boolean> = {};
+            var signatures = getSignaturesOfSymbol(symbol);
+
+            for (var i = 0; i < signatures.length; i++) {
+                var signature1 = signatures[i];
+                for (var j = 0; j < i; j++) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: So I have strong reservations about this. In `lib.d.ts` we have 122 overloads for `addEventListener`. As an `O(n^2)` operation, this is clearly expensive (14884 iterations).

Two things I could think of:
* If signature declarations could be totally ordered, one could "sort" the list of signatures, then pass through and find runs of duplicates (`O(n log n)` - or ~845 iterations for `addEventListener`)
* Find a way to serialize each signature declaration along with its parent, put each declaration into a map if it doesn't currently exist; if it does, report an error. (`O(n) assuming no background collisions, 122 iterations for `addEventListener`, but serialization could be slow).

@ahejl

**Comment by mhegazy** [comment]
`o(n^2)` would not be feasible. closing. please resubmit if you have a better solution,

**Pull request closed without merging**